### PR TITLE
mma: add motor mixing algorithm

### PR DESCRIPTION
**Pull request by HBuczynski**
Add basic implementation of MMA.

PP-24

### CODE REVIEW

File: quadcontrol/mma.c [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	}
+}
+
+
+void mma_stop(void)
+{
+	unsigned int i;
+
+	for (i = 0; i < NUMBER_MOTORS; ++i) {
+		fputc(100000, mma_common.files[i]);
```
Comment (agkaminski) [inline quadcontrol/mma.c:56]: Maybe #define min/max/scaler PWM constants?

File: quadcontrol/mma.c [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	}
+}
+
+void mma_done(void)
+{
+	unsigned int i;
+
+	for(i = 0; i < NUMBER_MOTORS; ++i) {
```
Comment (agkaminski) [inline quadcontrol/mma.c:64]:
```suggestion
	for (i = 0; i < NUMBER_MOTORS; ++i) {
```

File: quadcontrol/mma.c [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	unsigned int i;
+	uint32_t pwm[NUMBER_MOTORS];
+
+	pwm[0] = palt + pyaw + ppitch + proll;
+	pwm[1] = palt - pyaw + ppitch - proll;
+	pwm[2] = palt - pyaw - ppitch + proll;
+	pwm[3] = palt + pyaw - ppitch - proll;
+
+	for (i = 0; i < NUMBER_MOTORS; ++i) {
+		if (pwm[i] > 1.0f) {
```
Comment (agkaminski) [inline quadcontrol/mma.c:41]: Maybe also check for pwm[i] >= 0?

File: quadcontrol/mma.c (hunk 2 above)
Comment (HBuczynski) [reply quadcontrol/mma.c]: Done

File: quadcontrol/mma.c [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+/* TODO: hardcoded assumption that motors are connected to PWM0...PWM3
+         PWM definitions should be moved to board_config.h */
+int mma_init(const quad_coeffs_t *coeffs)
+{
+	int err;
+	unsigned int i;
+	char buff[32];
+
+	for(i = 0; i < NUMBER_MOTORS; ++i) {
```
[the remark this answers is not in the document]
Comment (HBuczynski) [reply quadcontrol/mma.c]: Done

File: quadcontrol/mma.c (hunk 1 above)
Comment (HBuczynski) [reply quadcontrol/mma.c]: I have added `PWM_MIN_SCALER`. However I don't see a point to add a max value.

File: quadcontrol/mma.c [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+struct {
+	quad_coeffs_t coeffs;
+	FILE *files[NUMBER_MOTORS];
+} mma_common;
+
+
+void mma_control(float palt, float proll, float ppitch, float pyaw)
+{
+	unsigned int i;
+	uint32_t pwm[NUMBER_MOTORS];
```
[the remark this answers is not in the document]
Comment (HBuczynski) [reply quadcontrol/mma.c]: Done

File: quadcontrol/mma.c [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	for (i = 0; i < NUMBER_MOTORS; ++i) {
+		err = snprintf(buff, sizeof(buff), "/dev/pwm%u", i);
+		if (err >= sizeof(buff)) {
+			return -EINVAL;
+		}
+
+		mma_common.files[i] = fopen(buff, "w");
+		if (mma_common.files[i] == NULL) {
+			fprintf(stderr, "mma: cannot open pwm file\n");
+			return -1;
```
Comment (agkaminski) [inline quadcontrol/mma.c:94]: Nitpick - potential fd leak on both returns

**Review by agkaminski — APPROVED** [review]
Code and logic LGTM, fd leak not a merge blocker

File: quadcontrol/mma.c [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+	for (i = 0; i < NUMBER_MOTORS; ++i) {
+		if (pwm[i] > 1.0f) {
+			pwm[i] = 1.0f;
+		}
+		else if (pwm[i] < 0) {
+			pwm[i] = 0.0f;
+		}
+
+		tmp = (uint32_t)(pwm[i] * 100.0f * 1000.0f + (float)PWM_MIN_SCALER);
```
[the remark this answers is not in the document]
Comment (HBuczynski) [reply quadcontrol/mma.c]: Yeah, I agree with You.

File: quadcontrol/mma.c (hunk 6 above)
Comment (HBuczynski) [reply quadcontrol/mma.c]: I have added `mma_done()` before returning with error.

**Pull request merged**